### PR TITLE
ability to specify comments and colours for permission rules

### PR DESCRIPTION
**Pull request by Guite**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| New feature?      | yes
| BC breaks?        | no
| Deprecations?     | no
| Fixed tickets     | #914 
| Refs tickets      | -
| License           | MIT
| Changelog updated | yes

## Description

This PR introduces two new fields for permission rules:

- an optional comment which is shown as tooltip of the corresponding table row
- an optional colour which corresponds to a Bootstrap context class (info, danger, etc.)

It is also possible to filter by colour in the same way as you can filter by group and/or component.

While the two new fields are added to the permission entity the old `realm` field is removed in the same step.

### CODE REVIEW

File: src/system/PermissionsModule/Listener/Core3UpgradeListener.php [hunk 1]
```diff
@@ -34,8 +34,20 @@ public function upgrade(GenericEvent $event): void
...
+        if ($this->columnExists('group_perms', 'realm')) {
+            $sql = 'ALTER TABLE `group_perms` DROP COLUMN `realm`;';
+            $this->conn->executeQuery($sql);
+        }
         if ($this->columnExists('group_perms', 'bond')) {
-            $sql = 'ALTER TABLE `group_perms` DROP COLUMN `bond`';
+            $sql = 'ALTER TABLE `group_perms` DROP COLUMN `bond`;';
+            $this->conn->executeQuery($sql);
+        }
+        if ($this->columnExists('group_perms', 'comment')) {
```
Comment (craigh) [inline src/system/PermissionsModule/Listener/Core3UpgradeListener.php]: this won't work like you intend. you are basically saying, 'if the column exists, then add it'

Comment (Guite) [reply src/system/PermissionsModule/Listener/Core3UpgradeListener.php]: oops, missing `!`

File: src/system/ThemeModule/EventListener/Core3UpgradeListener.php [hunk 2]
```diff
@@ -35,7 +35,7 @@ public function upgrade(GenericEvent $event): void
             return;
         }
         if ($this->columnExists('themes', 'xhtml')) {
-            $sql = 'ALTER TABLE `themes` DROP COLUMN `xhtml`';
+            $sql = 'ALTER TABLE `themes` DROP COLUMN `xhtml`;';
```
Comment (craigh) [inline src/system/ThemeModule/EventListener/Core3UpgradeListener.php]: please remove this. it is unrelated to this PR and the theme table will be dropped by me anyway.

File: src/system/PermissionsModule/Listener/Core3UpgradeListener.php (hunk 1 above)
Comment (Guite) [reply src/system/PermissionsModule/Listener/Core3UpgradeListener.php]: fixed, thanks

File: src/system/ThemeModule/EventListener/Core3UpgradeListener.php (hunk 2 above)
Comment (Guite) [reply src/system/ThemeModule/EventListener/Core3UpgradeListener.php]: done

File: src/system/PermissionsModule/Tests/Api/Fixtures/StubPermissionRepository.php [hunk 3]
```diff
@@ -100,6 +95,12 @@ public function getAllComponents(): array
         return [];
     }
 
+    public function getAllColours(): array
+    {
+        // TODO: Implement getAllColours() method.
```
Comment (craigh) [inline src/system/PermissionsModule/Tests/Api/Fixtures/StubPermissionRepository.php]: ?
either implement or remove comment

Comment (Guite) [reply src/system/PermissionsModule/Tests/Api/Fixtures/StubPermissionRepository.php]: the whole stub is full of these comments, I just copied `getAllComponents`

Comment (craigh) [reply src/system/PermissionsModule/Tests/Api/Fixtures/StubPermissionRepository.php]: ok. will you just remove all the comments then? 😄

Comment (Guite) [reply src/system/PermissionsModule/Tests/Api/Fixtures/StubPermissionRepository.php]: yes; done

**Pull request merged**